syntax = "proto3";

package cs739;


service RBS {
  rpc CheckPrimary(EmptyPacket) returns(Response);
  rpc Read(ReadRequest) returns(Response);
  rpc Write(WriteRequest) returns(Response);
}

service RaftInterface {
  rpc RequestVote (RequestVoteRequest) returns (RequestVoteResponse) {}
  rpc AppendEntries (AppendEntriesRequest) returns (AppendEntriesResponse) {}
}

message RequestVoteRequest {
  int64 term = 1;
  int64 candidate_id = 2;
  int64 last_log_index = 3;
  int64 last_log_term = 4;
}

message RequestVoteResponse {
  int64 term = 1;
  bool vote_granted = 2;
}

message AppendEntriesRequest {
  int64 term = 1;
  int64 leader_id = 2;
  int64 prev_log_index = 3;
  int64 prev_log_term = 4;
  repeated Entry entries = 5;
  int64 leader_commit = 6;
}

message AppendEntriesResponse {
  int64 term = 1;
  bool success = 2;
}

message Entry {
    int64 term = 1;
    int64 address = 2;
    bytes data = 3;
}

message ReadRequest {
  int64 address = 1;
}

message WriteRequest {
  int64 address = 1;
  bytes data = 2;
}

message Response {
  int32 return_code = 1;
  int32 error_code = 2;
  int64 current_leader = 3;
  int32 primary = 4;
  bytes data = 5;
}

message EmptyPacket {}
